lib/semantics/io.cc

```diff
@@ -0,0 +1,619 @@
+  case ParseKind::Encoding: specKind = SpecifierKind::Encoding; break;
+  case ParseKind::Form: specKind = SpecifierKind::Form; break;
+  case ParseKind::Pad: specKind = SpecifierKind::Pad; break;
+  case ParseKind::Position: specKind = SpecifierKind::Position; break;
+  case ParseKind::Round: specKind = SpecifierKind::Round; break;
+  case ParseKind::Sign: specKind = SpecifierKind::Sign; break;
+  case ParseKind::Convert: specKind = SpecifierKind::Convert; break;
+  case ParseKind::Dispose: specKind = SpecifierKind::Dispose; break;
+  }
+  SetSpecifier(specKind);
+  if (const auto *expr{GetExpr(std::get<parser::ScalarDefaultCharExpr>(spec.t)
+                                   .thing.thing.value())}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    if (const DefaultCharConstantType *
+        charConst{evaluate::UnwrapExpr<DefaultCharConstantType>(foldExpr)}) {
+      std::string s{parser::ToUpperCaseLetters(**charConst)};
+      if (specKind == SpecifierKind::Access) {
+        hasKnownAccess_ = true;
+        hasAccessDirect_ = s == "DIRECT"s;
+        hasAccessStream_ = s == "STREAM"s;
+      }
+      CheckStringValue(specKind, **charConst, parser::FindSourceLocation(spec));
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Newunit &) {
+  SetSpecifier(SpecifierKind::Newunit);
+}
+
+void IoChecker::Enter(const parser::ConnectSpec::Recl &spec) {
+  SetSpecifier(SpecifierKind::Recl);
+  if (const auto *expr{GetExpr(spec)}) {
+    const auto &foldExpr{
+        evaluate::Fold(context_.foldingContext(), common::Clone(*expr))};
+    const std::optional<std::int64_t> &recl{evaluate::ToInt64(foldExpr)};
+    if (recl && *recl <= 0) {
+      context_.Say(parser::FindSourceLocation(spec),
+          "RECL value (%d) must be positive"_err_en_US,
+          std::move(*recl));  // 12.5.6.15
+    }
+  }
+}
+
+void IoChecker::Enter(const parser::EndLabel &spec) {
+  SetSpecifier(SpecifierKind::End);
+}
+
+void IoChecker::Enter(const parser::EorLabel &spec) {
+  SetSpecifier(SpecifierKind::Eor);
+}
+
+void IoChecker::Enter(const parser::ErrLabel &spec) {
+  SetSpecifier(SpecifierKind::Err);
+}
+
+void IoChecker::Enter(const parser::FileUnitNumber &spec) {
+  SetSpecifier(SpecifierKind::Unit);
+  hasNumberUnit_ = true;
+}
+
+void IoChecker::Enter(const parser::Format &spec) {
+  SetSpecifier(SpecifierKind::Fmt);
+  if (std::get_if<parser::Star>(&spec.u)) {
+    hasStarFormat_ = true;
+  } else if (std::get_if<parser::Label>(&spec.u)) {
+    // Format statement format should be validated elsewhere.
+    hasLabelFormat_ = true;
+  } else {
+    hasCharFormat_ = true;
+    // TODO: validate compile-time constant format -- 12.6.2.2
+  }
+}
+
+void IoChecker::Enter(const parser::IdExpr &spec) {
+  SetSpecifier(SpecifierKind::Id);
+}
+
+void IoChecker::Enter(const parser::IdVariable &spec) {
+  SetSpecifier(SpecifierKind::Id);
+  auto expr{AnalyzeExpr(context_, spec.v)};
```

Comment:
You shouldn't need to call AnalyzeExpr, just `GetExpr(spec)` should work. But you do need to check it returns a valid result, there could have been an error.

Reply:
Done